src/parquet/ReadParquet.cpp

```diff
@@ -1,5 +1,407 @@
+    ReaderType* reader = static_cast<ReaderType*>(column_reader.get());
+    *startIdx -= reader->Skip(*startIdx);
+
+    // TODO find better variable names. values_read and num_read are confusing
+    int64_t values_read = 0;
+
+    int64_t num_read = 0;
+    if (nullMode == noNulls) {
+      while (reader->HasNext() && row_idx < numElems) {
+        PqType* tmpArr = (PqType*)malloc(batchSize * sizeof(PqType));
+        if((numElems - row_idx) < batchSize) // adjust batchSize if needed
+          batchSize = numElems - row_idx;
+        reader->ReadBatch(batchSize, nullptr, nullptr, tmpArr, &values_read);
+
+        // promote to larger type
+        for (int64_t j = 0; j < values_read; j++) {
+          chpl_ptr[row_idx+j] = (ChplType)tmpArr[j];
+        }
+
+        row_idx+=values_read;
+        num_read+=values_read;
+      }
+    }
+    else {
+      int16_t definition_level; // nullable type and only reading single records in batch
+      while (reader->HasNext() && row_idx < numElems) {
+        PqType value = 0;
+        auto ret = reader->ReadBatch(1, &definition_level, nullptr, &value,
+                                     &values_read);
+        // if values_read is 0, that means that it was a null value
+        if(values_read > 0) {
+          chpl_ptr[row_idx] = (ChplType) value;
+        }
+        else {
+          chpl_ptr[row_idx] = NAN;
+        }
+        row_idx++;
+        num_read++;
+      }
+    }
+    return num_read;
+  }
+
+  template<> int64_t read<arrow::DoubleType>() {
+    using Types = TypeBundle<arrow::DoubleType>;
+    using ReaderType = typename Types::ReaderType;
+    using ChplType = typename Types::ChplType;
+    using PqType = typename Types::PqType;
+
+    auto chpl_ptr = (ChplType*)chpl_arr;
+    ReaderType* reader = static_cast<ReaderType*>(column_reader.get());
+    *startIdx -= reader->Skip(*startIdx);
+
+    // TODO find better variable names. values_read and num_read are confusing
+    int64_t values_read = 0;
+
+    int64_t num_read = 0;
+    if (nullMode == noNulls) {
+      while (reader->HasNext() && row_idx < numElems) {
+        if((numElems - row_idx) < batchSize) // adjust batchSize if needed
+          batchSize = numElems - row_idx;
+        reader->ReadBatch(batchSize, nullptr, nullptr, &chpl_ptr[row_idx],
+                          &values_read);
+        row_idx+=values_read;
+        num_read+=values_read;
+      }
+    }
+    else {
+      int16_t definition_level; // nullable type and only reading single records in batch
+      while (reader->HasNext() && row_idx < numElems) {
+        PqType value = 0;
+        auto ret = reader->ReadBatch(1, &definition_level, nullptr, &value,
+                                     &values_read);
+        // if values_read is 0, that means that it was a null value
+        if(values_read > 0) {
+          chpl_ptr[row_idx] = (ChplType) value;
+        }
+        else {
+          chpl_ptr[row_idx] = NAN;
+        }
```

Comment:
same comments as about this block

Reply:
Considering leaving this as-is as per my previous comment.